### PR TITLE
feat/rework: port to app-platform

### PR DESCRIPTION
**Pull request by awgaan**
Some *Schema* logic shared with master but otherwise a complete rewrite.

### CODE REVIEW

**Comment by varl** [comment]
Very cool, @awgaan! Looks and feels much simpler than the existing code base.

Do you know which commit you forked from? We need to make sure that anything added to `master` after that point is represented in this fork. It looks like it is this one: https://github.com/dhis2/import-export-app/commits/a7c52cfea5b5815f7bd92df98da1abfcf031518e

```
53605898c1e737696515a020930256a21d11c9d4 chore: fix linting errors
014cd558a745188cd4351225924f1f62c839e98e refactor(reducers): remove action default value (empty object)
eda7aafb8520e745f93ae72af05cb514bec355c6 refactor(radio group): remove unused rest variable
95fad94c8c13c5112495ef205d771b1003362817 chore(jest): remove skip preflight check dev env var
bd9ded0558cf160ec3705ce68fb42b883f457096 test: update and add component jest tests
4f76870586baf513e59d5789d6115f1b9a3f66bc fix: adjust tests to new style and fix found refactoring bugs
4bb0762e7249021b2be63f958cdbe8418f96eebf refactor(metadata dependency export): use RFF
be4fe6d06059f99581d156eea6c2ae9a46646ee6 feat(object): add api helpers
a0cc1bc284f92f2f4780345c4a4891f5e0056e89 feat(object): add redux components
39c0ec59acfc259c40a0b10b91e7fb492bb5ddd8 feat(object): add ObjectList and ObjectType components
f0a3baf7f98d3a0e9995d2cda4c098f6dae6504c refactor(event export page): use RFF
6c9a8c62350038f5035c2bdd75a6a4c3d5b8263f feat(program & stages): add translations
c0cd92f75421ce27ec1da46997d72ae349cc2a97 feat(input components): add missing components for Event export page
3459942c5f7c203388a1b2e7968dd452ca90d357 feat(program stages): add redux components
c55c64c4ac6354f43a991ec7144dbc47e33ebbc5 feat(program): add redux components
7412dff5c05960ce5babde9e1d14957c10e641b4 fix(api helper): extract correct name in getProgramStages helper
7ce3b4c827cc2620a1563b5feb49e476f8b2e5ef refactor(select): allow nullable
21be410cded0eef76a85510b0ffadeefcd13235d refactor(org unit tree): accept multiple prop
0b5f920f945c56f7ad0f57ada8a55e78d69ae4e5 refactor(org unit tree): move state to redux
702292a38142990b9c7eee4bc6382a4b8f08e347 chore: update yarn.lock and prevent CRA warnings
389a63283e1c8991e403c7223853d1dc3514b179 refactor(data export): use RFF and new components
060a6f82679285d7c8a29daaad4c195feef7e58f refactor(data import): remove loading data, done by inputs themselves
13a3d4667693ac5ecec9206d30f78abf5b08d861 feat: add fields used for data export page
741b359e14141fce0041a6b22f88fecdfe3a75a7 refactor: make inputs autonomous & their styles consistent
1d99b4378c7abd712a6e179399c6bd1cc33b7517 chore(deps): remove unused url-join dependency
7c5b8067ce919ac0ca23801953752f8ad52f7be5 chore(translations): sync French translations from transifex [skip ci] (master)
55a62b8f25fd23e6014c6a5e791b3c8e0c083bdf test(pages): check import & export endpoint params
9501325214a58e65772c30bc75d3e1c55ee79595 refactor(input components): add data-test props & small adjustments
7628b4c0964b5fd51d23c3ddda165c1dae6e0bd9 fix(import meta data): send correct file format
c66e9e12606fcd4a38cbbdc2ae1bc1ca48db5125 refactor: replace data test id with data test (#311)
836f0f619d2c6fac3ba6a8efee95755179512eef chore(features): add missing "unique" description
```

Some of these will be relevant, and some will not. For example, introducing the Cypress tests are must-have as they will allow a smoother regression test of this branch.

**Comment by awgaan** [comment]
@varl yup, that's the one. I will take a look at the current cypress tests and try to replicate them for this branch.

**Comment by Mohammer5** [comment]
@awgaan Hi, thanks for putting so much work into this!

Right now it's quite hard to review this PR, it still needs a lot of refactoring in terms of cleaning up.
We're trying to transition to react-final-forms and soon we'll be able to use our new `ui-forms` library to assist us.

I'd propose you implement react-final-form (you can have a look at the current implementation on the master branch how [we've done it in a recent refactor](https://github.com/dhis2/import-export-app/tree/master/src/components/Inputs)).

I'd also suggest to put the loading logic to the components that use certain data, so it's easy to find code that belongs together. The components and their contents in the `src/components/Inputs` folder are quite solid already I think, so that's def. something that you could copy / use as cheat sheet.

Once that's done, it'll be much easier to review this amount of code.

**Comment by awgaan** [comment]
> @awgaan Hi, thanks for putting so much work into this!
> 
> Right now it's quite hard to review this PR, it still needs a lot of refactoring in terms of cleaning up.
> We're trying to transition to react-final-forms and soon we'll be able to use our new `ui-forms` library to assist us.
>
> I'd propose you implement react-final-form (you can have a look at the current implementation on the master branch how [we've done it in a recent refactor](https://github.com/dhis2/import-export-app/tree/master/src/components/Inputs)).

I was under the impression that react-final-forms was made for large complex forms. The import-export app has at most, what, 8-10 fields. Are you saying that we should use react-final-forms throughout all apps for uniformity?

> I'd also suggest to put the loading logic to the components that use certain data, so it's easy to find code that belongs together. The components and their contents in the `src/components/Inputs` folder are quite solid already I think, so that's def. something that you could copy / use as cheat sheet.

Are you referring to the use of "loading" in the import pages? I think that particular loading state belongs on that level since it shows the progress of the file upload.

> Once that's done, it'll be much easier to review this amount of code.

Thank you for the feedback ;)

**Comment by Mohammer5** [comment]
> I was under the impression that react-final-forms was made for large complex forms.

React-final-form definitely has some functionality to make large forms work smoothly, but I wouldn't say it's exclusively for large forms.

> Are you saying that we should use react-final-forms throughout all apps for uniformity?

Consistency across the projects is definitely a good thing, but we also can leverage on react-final-form's functionality (abstracting form state, validation and error handling) and use our (soon to be released) `ui-forms` library

<hr />

> Are you referring to the use of "loading" in the import pages? I think that particular loading state belongs on that level since it shows the progress of the file upload.

No, I was referring to loading programs, data sets, organisation units, schemas, etc.
But the way I said it maybe wasn't exactly what I meant either, so good you're bringing this up.

This is from the `src/App.js`
```js
useEffect(() => {
    fetchAndSetAttributes({
        apiBaseUrl: `${baseUrl}/api/`,
        setDataElementIdSchemeOptions,
        setOrgUnitIdSchemeOptions,
    })
}, [])
```

As the data element id scheme is directly connected to the response of the endpoint call with this filter: `unique:eq:true&filter=${attr}:eq:true`, the trigger to fetch these attributes should be close to the data element id scheme, same goes for the org unit id scheme.

In the current implementation of the import-export-app, everything related to the data element is scheme (key, default value, default options, label, logic when to load attributes & the component itself) is inside [the data element is scheme file](https://github.com/dhis2/import-export-app/blob/master/src/components/Inputs/DataElementIdScheme.js), but the actual fetching and state logic is abstracted.

<hr />

> Are you referring to the use of "loading" in the import pages? I think that particular loading state belongs on that level since it shows the progress of the file upload.

I'm not exactly sure though what you mean by '"loading" in the import pages' though. Are you talking about the import progress and task log?

There's a lot of "task" related code in the `src/App.js` which should go into its own files as well. Currently it's very hard to follow what's happening in the App component, the functions `addTask`, `fetchEvent` and `fetchSummary` should go to their own files and/or folder.

And that's what I meant with "it still needs a lot of refactoring in terms of cleaning up". Many functions do many things and could be split up (e. g. the `fetchAndSetAttributes` helper).

**Comment by Mohammer5** [comment]
Btw, this PR is still in draft mode. Can you lay out what you want to do / needs to be done before you feel ready to change this one to be ready for review?
No need to talk about stuff you want to change anyways.

**Comment by awgaan** [comment]
Completely agree with most of what you've written.
Today's commits move task/job logic to own component.
Scheme components now fetch their own data.
`App.js` lost a lot of weight.

**Comment by awgaan** [comment]
> Btw, this PR is still in draft mode. Can you lay out what you want to do / needs to be done before you feel ready to change this one to be ready for review?
> No need to talk about stuff you want to change anyways.

Needs to be done before review:
- Component unit tests
- Cypress E2E tests

**Comment by awgaan** [comment]
All previous Cypress tests are passing now.
What more do I need?
I feel like I am relying too heavily on snapshot tests a couple of places where more in depth tests might've been more appropriate. But those are covered by the E2E tests anyway.

**Comment by varl** [comment]
Good job on this port. Very cool! :tada: :muscle: 

We are a bit too close to the 2.34 cutoff, so we should defer merging this until after 2.34 has been released. That gives QA more time to reg test the application as well.

**Comment by varl** [comment]
@awgaan Could you rebase this on master and make sure that all changes since last time it was synced are covered? I don't think that there will be many changes to master from today, and as soon as 2.34 is branched out it would be cool to merge this in.

**Comment by awgaan** [comment]
Will do.

**Comment by awgaan** [comment]
I went through all the changes and made sure they were present on this branch.
Also rebased.

**Comment by varl** [comment]
2.34 has been branched, we can get this into `master` now and work on getting it ready for 2.35!

**Comment by awgaan** [comment]
@Mohammer5 That's fantastic! I'll get right to it.

**Comment by Mohammer5** [comment]
@awgaan Ah just saw that there is an editorconfig.. Weird that I didn't see it and it wasn't applied in my editor.. Sorry about that..

Btw, none of what I've written above is like a shove-it-down-your-throat comment, we can discuss every single comment you want to talk about or have a different opinion.

**Comment by awgaan** [comment]
## .editorconfig
Added

https://github.com/dhis2/import-export-app/pull/385/commits/7c5bd36835a39b6cf7af17253caa548c5bcadf7c

## Default values for state and/or variables
Completely agree with you here.
Should be okay now.

## Creating functions inside components
Converted several components and functions to hooks.
Cleaned up wherever I saw fit.
Let me know if I missed something.

## `data-test` ids
I am not sure I agree. I find having a central file where all test-ids are stored by category and component quite convenient.
I left it as it were, but if you or the rest of the team prefer having the ids right there in the code, I will of course respect that.

## Abbreviations
You're right. Abbreviations aren't useful anymore. I changed that specific import (remnant from current code) to be more describing.

https://github.com/dhis2/import-export-app/pull/385/commits/bc235ddc6f964bfac9e40c2c9fe77363efd00836

## `useDataQuery` hook
`onComplete` and `onError` functions are used if I need to perform some operation on the received data before using it. Otherwise I switched to using `{error, data}` returned from the hook.

## Empty .module.css files & import comments
Removed

https://github.com/dhis2/import-export-app/pull/385/commits/543b402f9ee0486651b73765502c9cf32364f537

## Async/Await
Cleaned up unnecessary awaits.

## WithSchemes
Removed.
Still uses `fetch` because the `app-runtime` doesn't support multiple filters yet.
Waiting for [https://github.com/dhis2/app-runtime/pull/261](https://github.com/dhis2/app-runtime/pull/261) to merge.

https://github.com/dhis2/import-export-app/pull/385/commits/692eb6b0f4e7ec7b99944d0292dec7c0c8e2a49c

https://github.com/dhis2/import-export-app/pull/385/commits/2b6e6cbf4b307c2f88bef3c506c89b7e4c2f729a

## WithTasks
Converted to a hook and used directly in `App`.

https://github.com/dhis2/import-export-app/pull/385/commits/8b70e307ee6201872da6dd7559051cb537860738

## WithUser
Converted to a hook and used directly in `App`.

https://github.com/dhis2/import-export-app/pull/385/commits/955f23aaa0a837db4f1b2ab992d511a1a93611be

## Skeleton
Removed.
Creater a separate `Router` component.
`App` now renders the `Sidebar` and `Router`.

https://github.com/dhis2/import-export-app/pull/385/commits/4e3054bc21abf57e60bc81ed1b2df6e8a6a9e74e

## `src/utils/pages`
I prefer as it  is because we avoid some repetition and I think most of the information in this file belongs together. Maybe let @varl have the final say?

## React-Final-Form
Not converted to an RFF app yet.
Soon. Maybe.

## Page components
### String conversion
Using `.toString` instead of string literals.

https://github.com/dhis2/import-export-app/pull/385/commits/3d89bd1a2121d77bf04784917349e4dffa00434d

### Initial state type / type safety
Using the solution you suggested.

### params in the onSubmit function
Moved static parameters to the top of the list.

### ProgramStageSelect, ObjectSelect, ClassKey
Converted to hooks.

https://github.com/dhis2/import-export-app/pull/385/commits/1189d1873dbb7e622bc583e7fc57e49a0101a8cc

https://github.com/dhis2/import-export-app/pull/385/commits/f889bfcb73b4a6fe4d609f4fdbd09e74537e6ff7

https://github.com/dhis2/import-export-app/pull/385/commits/b4c51b2843dd8bb6d73ad7cd1432001eb4abb67c

## Components
### DataSetPicker, OrgUnitTree, ProgramPicker, Schemas
JSX return statements changed to follow the convention you suggested.

https://github.com/dhis2/import-export-app/pull/385/commits/1d8eb1c5ecb953d6ea5d8b8964474489e8324d34

### FileUpload
OnChange function in-lined.

https://github.com/dhis2/import-export-app/pull/385/commits/dd016be62a2ff308a4f2a4af7a41cfe50b936111

### FormAlerts, FormField
Will be changed when switching over to RFF and `ui-forms`.

### Icon
Using the `svg` directly.

https://github.com/dhis2/import-export-app/pull/385/commits/ea708a02394dc60cf10d21bcb88132a79fb4a961

### JobOverview
Comments added.

https://github.com/dhis2/import-export-app/pull/385/commits/6e69d74a96ee0a9e3094d7756916360463fbd338

### JobSummary
Split into several smaller components.

https://github.com/dhis2/import-export-app/pull/385/commits/801d91ba3416fdab498c35925c96e1b3bc4efd72

### Tag
Waiting for `app-shell` to upgrade its `ui-core` dependency so we can use the official `Tag` component.

### Summary
Extracted statistics function.

https://github.com/dhis2/import-export-app/pull/385/commits/55dd935579be74c4fa12b072c65629e9fdb13733

**Comment by Mohammer5** [comment]
## data-test ids

> if you or the rest of the team prefer having the ids right there in the code, I will of course respect that.

No need to rush. We still have time to talk about this, 2.35 will not be released soon. Let's talk about it a bit more.

> I find having a central file . . . quite convenient.

Can you explain to me what the benefits are from your POV?

<hr />

I can add some points why I see the benefit in putting them into the files where they're used:

#### Benefit of config files

When having a couple of places (or more), a config file does make a lot of sense. It simplifies maintenance as changing config only needs to be done in one place and will affect multiple places. It ensures deterministic behavior as well as all code uses the same data.

#### Drawbacks of config files

There surely must be a reason why we're not using config files for everything. Config files add inconvenience when working on code that uses those config files: 1. Context switch: You don't know values directly, which means you have to make a small context switch (from module to config module and back), which is an interrupt of the brain's workflow, just to see the actual value. 2. Makes editing more complicated: You will have to handle multiple files at the same time when reading through code, which one the one side is another context switch, but also increases editing complexity (especially when handling multiple files already) as you would have to have at least one extra file open to see values

#### Using config files is more error prone than using the value directly

Of course there's a clear benefit when reducing repetition, but regardless of that, there's always the possibility that consumers of the config files have a typo or use the wrong value. In this case that would result in "undefined" which wouldn't trigger an error or warning, it's a bug until noticed.

#### `data-test` ids should be used only once

I assume that every data-test value in the config file is only used once, re-using them wouldn't make much sense and should cause a refactor. If there are several components which are mutually exclusive and share the same data-test id, their container should be extracted if they have common/the same attributes.

#### Summary of the above 

So although having data-test ids in a config file is not too much of a burden as their value is more or less irrelevant, it's not a counter-argument to what I've written above. They're not shared between multiple files, introduce a cognitive overhead, to get to know their value you have to go to more than the one file and while developing you have the slight overhead of having a wee bit of guesswork that the data-test config is correct and that the code is using the correct config.

## useDataQuery hook

In the following files `onError` can either be removed or replaced with the return value of the hook

* `src/hooks/useClassKeys.js`
* `src/hooks/useObject.js`
* `src/components/ProgramPicker/ProgramPicker.js`
* `src/components/OrgUnitTree/OrgUnitTree.js`
* `src/components/Schemas/Schemas.js`

## `src/utils/pages`

Again, we have time to talk about this, no need to rush.

> I think most of the information in this file belongs together

Can you explain that? I'm not sure I understand it correctly.

The data for an individual page belongs together, but it also belong to the page it's connected to. Moving it there wouldn't take them apart, they'd be still together.

If you're talking about that data from different pages belongs together, then I don't really see why. Sure, they have the same shape, but that's solely because you wrote it that way. We could argue the same for page/view components (same shape, same role) and put them all into the same file

And yes, we would have some repetition, but we'd also have more explicitness and quite a bit less complexity:

* We'd reduce 12 imports of that file to 3
* If I had to look for which icon/description/whatever is used of a particular page, I don't have to go to multiple files and I don't have to remember that there are config files. I simply go to the page file and get all the information.
* The Router can explicitly state which page is the default page (instead of relying on the order of `ImportPages`, which no one will know is the case).
* Looking at the `Router` gives a much better overview of what pages are available, as that's 1. a much likelier place to look for router configuration than a `utils` file and 2. is much easier to scan as the required information for a route is directly mapped to named exports from the page files
* The order in which to render and group pages is currently done through three constants: `ImportPages`, `ExportPages` & `Pages`. `Pages` is only used by the Router which I already argued against. The other two are used only by the `SideBar` component. So instead of creating multiple layers of configuration, this could be done statically in the `SideBar` component directly.
* Same arguments as I made for config files when talking about the `data-test` ids

## React-Final-Form

> Soon. Maybe.

What do you mean by this? If I recall correctly we agreed on using RFF during the meeting in Oslo. What's holding you off? And what is the meaning of `Maybe.`? You're unsure about using it?

**Comment by awgaan** [comment]
## data-test
The intention behind the config file was to avoid typos, since most of the components will complain if they get an undefined data-test. But in the end, I agree; `data-test` values feel more at home straight in the jsx.

4d0038b 

## useDataQuery error
- *useClassKeys*: error return value is already used. `onError` is used to log the error.
- *useObjects*: `query` is used within a `useEffect` and we have to define a mutable error variable in the _global_ scope and destructure the `engine.query` error return value into this variable. That feels awkward, and I left it as it was. Maybe you thought of a better way to handle this?
- *ProgramPicker*:  3f1dd2f
- *Schemas*: 3f1dd2f
- *OrgUnitTree*: error can occur in multiple places

## pages config file
I mostly wanted to avoid repetition. But in the end the amount of repeated code was acceptable. As a bonus the `Router` is much more readable.

bd7b996 

## React-Final-Form
I should've phrased that better. I wanted to resolve the current issues before I started on that. It's 100% my intention to transition over to RFF as soon as possible.

File: package.json [hunk 1]
```diff
@@ -1,84 +1,54 @@
...
+        "eslint-config-prettier": "^6.7.0",
+        "eslint-config-react-app": "^5.0.2",
+        "eslint-plugin-cypress": "^2.7.0",
+        "eslint-plugin-flowtype": "^4.5.2",
+        "eslint-plugin-import": "^2.18.2",
+        "eslint-plugin-jsx-a11y": "^6.2.3",
+        "eslint-plugin-prettier": "^3.1.1",
+        "eslint-plugin-react": "^7.17.0",
+        "eslint-plugin-react-hooks": "^2.3.0",
+        "prettier": "^1.19.1",
```
Comment (amcgee) [inline package.json]: Hey @awgaan - this looks really cool!  I think all these `babel` and `eslint` dependencies should be extraneous now, can we remove lines 28 to 42?

Comment (awgaan) [reply package.json]: Done. e66d3f8

**Comment by awgaan** [comment]
Now using @dhis2/ui-forms. All tests passing.

File: src/utils/xhr.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+const onReadyStateChange = ({ xhr, type, onResponse, onError }) => {
+    return async e => {
+        const status = Math.floor(xhr.status / 100)
+        if (xhr.readyState === 4 && status === 2) {
+            const idAndMsg = extractIdAndMessage(xhr)
+            onResponse({ ...idAndMsg, type: type })
+        } else if (xhr.readyState !== 4 && [3, 4, 5].includes(status)) {
+            onError(e)
+        }
```
[the remark this answers is not in the document]
Comment (awgaan) [reply src/utils/xhr.js]: https://github.com/dhis2/import-export-app/blob/223966838cf257f9042d1dbd14ba3ac6096a0275/src/utils/xhr.js#L33-L43

This should cover it, right? I think it's fair to wait for the xhr to complete before handling any errors.

[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/utils/xhr.js]: No, status can be less than 2 (i. e. when I tell the dev tools to block a certain endpoint/url, the status is 0.

The else if should be `else if (xhr.readyState === 4)`, which will cover all cases other than `status === 2`, no matter what the status is.

[the remark this answers is not in the document]
Comment (awgaan) [reply src/utils/xhr.js]: https://github.com/dhis2/import-export-app/blob/f28adf5f62eb0dd59387bcd970648953b35c27f6/src/utils/xhr.js#L39-L41

**Comment by varl** [comment]
@Mohammer5 @awgaan @amcgee 

I propose that we merge this into `master`, and move any remaining issues / refactors to JIRA issues that can be tackled in a more incremental manner so that they can be reviewed more easily. This would also allow others to contribute to the code base again and get comfortable with the new structure.

Since this is an app that is scheduled for 2.35 there is ample time to review the code and architecture.

**Comment by Mohammer5** [comment]
@varl agreed, I'll create Jira epic so we can collect the issues
@awgaan could you add Jira issues for all things that still need to be addressed?

**Comment by awgaan** [comment]
Added already existing relevant Jira issues to this epic.
Will create new issues if I manage to find some.
